### PR TITLE
time: do not panic on Interval::poll_tick if the sum of the timeout and period overflow

### PR DESCRIPTION
**Pull request by jxs**
## Motivation
when calling [tokio::time::interval(Duration::MAX)](https://docs.rs/tokio/latest/tokio/time/fn.interval.html) runtime panics with:
```
overflow when adding duration to instant
```
see reproduction [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cf23e831d76f4bd503f1a177b16a9a4a). Panic happens when adding `timeout` to to the `delay` deadline.

## Solution

This PR addresses the panic by calling [`Instant::checked_add`](https://doc.rust-lang.org/std/time/struct.Instant.html#method.checked_add) falling back to `Instant::far_future()` when it overflows. If needed I can add a test for it and update the documentation mentioning this behaviour.

Thanks!

### CODE REVIEW

**Comment by Darksonn** [comment]
`Instant::far_future()` is sufficient.

**Comment by wathenjiang** [comment]
Is it a good idea to add some explanation about this overflow behavior on the [`tokio:: time:: interval`](https://docs.rs/tokio/latest/tokio/time/fn.interval.html).

**Comment by Darksonn** [comment]
I don't think it's necessary, but it also doesn't hurt to mention that.

**Comment by jxs** [comment]
thanks Alice, added a test

File: tokio/src/time/interval.rs [hunk 1]
```diff
@@ -480,7 +480,9 @@ impl Interval {
             self.missed_tick_behavior
                 .next_timeout(timeout, now, self.period)
         } else {
-            timeout + self.period
+            timeout
+                .checked_add(self.period)
+                .unwrap_or(Instant::far_future())
```
[the remark this answers is not in the document]
Comment (jxs) [reply tokio/src/time/interval.rs]: Thanks Paolo, updated!

**Review by Darksonn — APPROVED** [review]
I've verified that the test does fail without this change. Thanks!

**Comment by Darksonn** [comment]
It seems there is a clippy failure.
```
error: redundant closure
   --> tokio/src/time/interval.rs:485:33
    |
485 |                 .unwrap_or_else(|| Instant::far_future())
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `Instant::far_future`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
    = note: `-D clippy::redundant-closure` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::redundant_closure)]`
```

**Comment by jxs** [comment]
Stress test is failing, but seems unrelated

**Comment by Darksonn** [comment]
Yes, see the CI issue above.

**Pull request merged**